gyronimo/dynamics/cartesian_boris.cc

```diff
@@ -29,91 +29,49 @@
-      metric_(
-          Bfield ? dynamic_cast<const metric_cartesian*>(Bfield->metric())
-                 : nullptr),
+      Oref_(B ? qom * codata::e / codata::m_proton * B->m_factor() * Tref_ : 1),
+      electric_field_(E), magnetic_field_(B),
+      iE_time_factor_(E ? Tref_ / E->t_factor() : 1),
+      iB_time_factor_(B ? Tref_ / B->t_factor() : 1),
+      metric_(B ? dynamic_cast<const metric_cartesian*>(B->metric()) : nullptr),
       my_morphism_(metric_ ? metric_->my_morphism() : nullptr) {
-  // test if fields exist
-  // if(!Efield) error(__func__, __FILE__, __LINE__,
-  // 							" Efield cannot be nullptr.",
-  // 1);
-  if (!Bfield)
-    error(__func__, __FILE__, __LINE__, " Bfield cannot be nullptr.", 1);
-
-  // test if fields are cartesian
+  if (!B) error(__func__, __FILE__, __LINE__, "no magnetic field.", 1);
+  if (E && B->metric() != E->metric())
+    error(__func__, __FILE__, __LINE__, "mismatched E/B coordinates.", 1);
   if (!metric_)
-    error(
-        __func__, __FILE__, __LINE__,
-        " Bfield must be in cartesian coordinates.", 1);
-  if (Efield) {
-    const metric_covariant* Emetric = Efield->metric();
-    if (Emetric != metric_)
-      error(
-          __func__, __FILE__, __LINE__,
-          " Efield must be in cartesian coordinates.", 1);
-  }
-
-  // test if normalization factors are consistent
-  if (Efield) {
-    double E_m_factor = Efield->m_factor();
-    double B_m_factor = Bfield->m_factor();
-    if (std::abs(1.0 - E_m_factor / (Vref_ * B_m_factor)) > 1.0e-14)
-      error(
-          __func__, __FILE__, __LINE__, " inconsistent Efield and Bfield.", 1);
-  }
+    error(__func__, __FILE__, __LINE__, "field has no metric_cartesian.", 1);
 }
 
 //! Performs a time step `dt` update to the state `in` and returns the result.
 cartesian_boris::state cartesian_boris::do_step(
-    const cartesian_boris::state& in, const double& time,
-    const double& dt) const {
-  // extract position and velocity from state
-  IR3 x_old = {in[0], in[1], in[2]};
-  IR3 v_old = {in[3], in[4], in[5]};
-
-  // calculate fields
-  IR3 Efield = {0.0, 0.0, 0.0};
-  if (electric_field_) {
-    double Etime = time * iEfield_time_factor_;
-    Efield = electric_field_->contravariant(x_old, Etime);
-  }
-  double Btime = time * iBfield_time_factor_;
-  double Bmag = magnetic_field_->magnitude(x_old, Btime);
-  IR3 Bversor = magnetic_field_->contravariant_versor(x_old, Btime);
-
-  // perform cartesian_boris step
-  IR3 v_new = electric_field_
-      ? boris_push(v_old, Oref_, Efield, Bmag, Bversor, dt)
-      : boris_push(v_old, Oref_, Bmag, Bversor, dt);
-  IR3 x_new = x_old + (Lref_ * dt) * v_new;
-
-  return {x_new[IR3::u], x_new[IR3::v], x_new[IR3::w],
-          v_new[IR3::u], v_new[IR3::v], v_new[IR3::w]};
+    const state& s, const double& time, const double& dt) const {
+  double Btime = time * iB_time_factor_;
+  const double tildeEref = Oref_ * iB_time_factor_ / (iE_time_factor_ * Vref_);
```

Comment:
Shouldn't it be: `tildeEref = Oref_ * electric_field_->m_factor() / (Vref_ * magnetic_field_->m_factor())`?
Also, since this is a constant, it can probably be calculated just in the constructor when `electric_field_` exists.

Reply:
Indeed a good candidate for the silliest of silly mistakes! Of course it needs to be corrected.